charms/layers/apache-nifi/reactive/apache_nifi.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (C) 2016  Qrama
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# pylint: disable=C0111, C0301
+
+import subprocess
+import tarfile
+
+from charms.reactive import when, when_all, when_not, set_state, remove_state
+from charmhelpers.core import hookenv
+from charmhelpers.core.host import service_restart, service_stop
+
+from jujubigdata.utils import re_edit_in_place
+
+
+@when_not('apache-nifi.installed')
+@when('java.installed')
+def install_apache_nifi():
+    conf = hookenv.config()
+    hookenv.log('Installing Apache NiFi')
+    hookenv.status_set('maintenance', 'Installing Apache NiFi')
+    tfile = tarfile.open(hookenv.resource_get('apache-nifi'), 'r')
+    filesdir = '%s/files' % hookenv.charm_dir()
+    tfile.extractall(filesdir)
+    re_edit_in_place('%s/nifi-1.0.0/conf/nifi.properties' % filesdir, {
+        r'.*nifi.web.http.port.*': 'nifi.web.http.port=%s' % conf['nifi-port'],
+    })
+    subprocess.check_call(['bash', '%s/nifi-1.0.0/bin/nifi.sh' % filesdir, 'install'])
+    service_restart('nifi')
+    hookenv.open_port(conf['nifi-port'])
+    hookenv.status_set('active', 'Apache NiFi has been installed and is running in standalone mode')
```

Comment:
Try to keep these messages as short as possible. Example: `Ready (standalone)`. Non-blocking but just something to keep in mind in the future.